### DISCUSSION TITLE
Physical units for physical lights

### DISCUSSION THREAD

**Discussion by no-lex** [project-imprimis/libprimis#176] [Ideas]
Some possiblities for making lights more physical (closer to reality) would be to use physical conventions for lights. Light "radius" isn't a concept in reality, for example; real lights are parameterized by their lumen output, and (ideally) radiate out to arbitrary distances.

Another idea for making lights more useful for artists would be to allow parameterization in terms of color temperature, which would make it easier to create natural-looking room lighting (as lightbulbs are usually intended to approximate incandescent color temperatures). This would make natural "warm" or "cool" lighting easy to make realisically.

Some food for thought:
How should light radii be cut off? (infinite radius lights would be rather expensive)
How would normalized colors best be represented? (if intensity is a parameter, we wouldn't want it to be double-settable)
How to handle realistic sunlight? Typical sunlight is ~1e5 lux, but is this a unit people recognize?

**Comment by no-lex** [comment]
More food for thought:

![falloff](https://user-images.githubusercontent.com/23445929/102701858-71239780-4210-11eb-9bde-cb30e257744d.png)

There's something somewhat fishy about how the light is falling off. Ideally, light from a point source drops off as 1/r^2 (in intensity). It doesn't appear that is happening here. The color of the terrain is `0x3E3E3E` at 1/2m and `0x3B3B3B` at 1m, a nearly imperceptible drop (really, the object at a distance of 1/2 meter should appear twice as bright).

**Comment by ghost** [comment]
**How should light radii be cut off?**
Cut off radius could be calculated based on some minimal light value (e.g. corresponding with changing the rendered value by 1, or some physical constant such as 1 lux), perhaps adjusted for environment (farthest wall in straight line or if it's going to be drowned out by other light sources, such as the sun).

**How would normalized colors best be represented?**
Perhaps it would be feasible to adopt the HSV color model, either dropping Value in favor of the light's brightness or using it as such directly (remembering about allowing overbright lights).

**How to handle realistic sunlight? Typical sunlight is ~1e5 lux, but is this a unit people recognize?**
Sunlight values may vary greatly depending on a number of factors in real life, however setting for a good default value of a clear sunny  day would be a good idea. 0.1Mlx sounds like a nice round value. Compare: https://en.wikipedia.org/wiki/Daylight


I applaud the idea of automating the creation of color temperatures. Perhaps allowing conversion for different systems such as temperature in Kelvin (remember that it also affects brightness!), HSV, RGB etc. to aid realistic lighting, then allowing modifying those values (e. g. setting the "base" temperature, adjusting the Green component, then the Value to simulate mercury vapor lamps) for greater artistic control.
